[jit] tracer flag to guard risky behaivors

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31860 [jit] Add dictionary as output of tracer
* **#36277 [jit] tracer flag to guard risky behaivors**
* #36411 Use unittest assertWarns instead

This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

Differential Revision: [D20998157](https://our.internmc.facebook.com/intern/diff/D20998157)

wanchaol committed:
[jit] tracer flag to guard risky behaivors

This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] tracer flag to guard risky behaivors"


This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

[ghstack-poisoned]

wanchaol commented on torch/onnx/utils.py:
```diff
@@ -288,7 +288,7 @@ def _trace(func, args, operator_export_type, return_outs=False):
 
     trace_graph, torch_out, inputs_states = \
-        torch.jit._get_trace_graph(func, args, _force_outplace=False, _return_inputs_states=True)
+        torch.jit._get_trace_graph(func, args, strict=False, _force_outplace=False, _return_inputs_states=True)
```
@houseroad I am relaxing the constraints of onnx exporter to use this `strict=False` flag, let me know if you see any issues from this.

jamesr66a commented on torch/testing/_internal/jit_utils.py:
```diff
@@ -527,6 +528,12 @@ def input_reduce(input, fn, acc):
 
         return ge
 
+    def checkTracerWarning(self, *args, **kwargs):
```
Instead of having this function, can we use `unittest.TestCase.assertWarnsRegex`? e.g. all existing callsites can be mechanically replaced like:

```
self.checkTracerWarning(foo, torch.rand(3, 4), _force_outplace=True)
```

to

```
with self.assertWarnsRegex(torch.jit.TracerWarning, "cause the trace to be incorrect""):
  torch.jit.trace(foo, torch.rand(3, 4), _force_outplace=True)
```

This will have the benefit that it's more explicit about what's going on and will force the writer to think about if there is a more specific string they could match on, helping to alleviate the risk of false-positives

jamesr66a commented on test/test_jit.py:
```diff
@@ -4300,6 +4300,26 @@ def forward(self, x):
+        with self.assertRaisesRegex(RuntimeError, r"Only tensors.+can be output from traced functions"):
+            traced_f = torch.jit.trace(f, [])
+
+
```
nit: extra newline

jamesr66a commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -857,7 +864,12 @@ const char* WARN_RESIZE =
+    "structure (i.e. dict keys) does not change based on the module's inputs, consider "
+    "using constant container structure instead(i.e. Tuple, NamedTuple) to avoid this "
+    "warning/error, If you absolutely need this and know the side effects, pass "
+    "strict=False to trace() to allow this behavior.";
```
This message talks exclusively about `dict`, but it is used when tracing a `list` output. Can we generalize this? Maybe something like:

```
" might cause the trace to be incorrect. This is only valid of the container structure does not change based on the module's inputs. Consider using a constant container instead (e.g. for `list`, use a `tuple` instead. for `dict`, use a `NamedTuple` instead). <...>
```

jamesr66a commented on torch/jit/__init__.py:
```diff
@@ -237,7 +237,7 @@ def export_opnames(m):
 
-def _get_trace_graph(f, args=(), kwargs=None, _force_outplace=False,
+def _get_trace_graph(f, args=(), kwargs=None, strict=True, _force_outplace=False,
                      return_inputs=False, _return_inputs_states=False):
```
Nothing actionable on this PR, but we'll have to keep an eye on the proliferation of boolean flags. Four boolean flags means 16 different "modes" the code can go through, and it's unlikely that we're testing all 16. I wonder if we can coalesce some of the other flags into one (e.g. an `onnx` flag or something)

jamesr66a reviewed (commented):
Great, thanks! Just some small comments. Functionality looks good

wanchaol committed:
Update on "[jit] tracer flag to guard risky behaivors"


This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] tracer flag to guard risky behaivors"


This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] tracer flag to guard risky behaivors"


This PR introduce a flag to the tracer that guard the risky behaviors
like adding list/dict as output of the tracer. Currently to ensure not
BC breaking user, we throw warning if the tracer output is list, and
will throw error when the tracer output is dict to enforce using this
flag (next PR)

[ghstack-poisoned]
